File: rex/test/modula/modula.m
GLOBAL {
(* /*
# define yySetPosition Attribute.Position.Line   := yyLineCount; \
  Attribute.Position.Column := yyChBufferIndex - yyLineStart - TokenLength; \
  IO.WriteI (IO.StdOutput, yyOffset (TokenIndex), 0); IO.WriteNl (IO.StdOutput);
*/ *)

PROCEDURE ErrorAttribute (Token: INTEGER; VAR Attribute: tScanAttribute);
   BEGIN
   END ErrorAttribute;
}

DEFAULT {
   Position.WritePosition (IO.StdOutput, Attribute.Position);
   IO.WriteS (IO.StdOutput, ": illegal character: ");
   yyEcho;
   IO.WriteNl (IO.StdOutput);
}

EOF {
IF yyStartState = Comment THEN
   Position.WritePosition (IO.StdOutput, Attribute.Position);
   IO.WriteS (IO.StdOutput, ": unclosed comment");
   IO.WriteNl (IO.StdOutput);
END;
}

DEFINE
   digit	= {0-9}		.
   letter	= {a-z A-Z}	.
   CmtCh	= - {*(\t\n}	.

START Comment

RULE

#STD, Comment# "(*"	:- {yyPush (Comment);}
#Comment#  "*)"		:- {yyPop;}
#Comment#  "(" | "*" | CmtCh + :- {}

#STD# digit +		,
#STD# digit + / ".."	: {RETURN 1;}
#STD# {0-7} + B		: {RETURN 2;}
#STD# {0-7} + C		: {RETURN 3;}
#STD# digit {0-9 A-F} * H : {RETURN 4;}
#STD# digit + "." digit * (E {+\-} ? digit +) ?	: {RETURN 5;}

#STD# ' - {\n'} * '	|
      \" - {\n"} * \"	: {RETURN 6;}

#STD# "#"		: {RETURN 7;}
#STD# "&"		: {RETURN 8;}
#STD# "("		: {RETURN 9;}
#STD# ")"		: {RETURN 10;}
#STD# "*"		: {RETURN 11;}
#STD# "+"		: {RETURN 12;}
#STD# ","		: {RETURN 13;}
#STD# "-"		: {RETURN 14;}
#STD# "."		: {RETURN 15;}
#STD# ".."		: {RETURN 16;}
#STD# "/"		: {RETURN 17;}
#STD# ":"		: {RETURN 18;}
#STD# ":="		: {RETURN 19;}
#STD# ";"		: {RETURN 20;}
#STD# "<"		: {RETURN 21;}
#STD# "<="		: {RETURN 22;}
#STD# "<>"		: {RETURN 23;}
#STD# "="		: {RETURN 24;}
#STD# ">"		: {RETURN 25;}
#STD# ">="		: {RETURN 26;}
#STD# "["		: {RETURN 27;}
#STD# "]"		: {RETURN 28;}
#STD# "^"		: {RETURN 29;}
#STD# "{"		: {RETURN 30;}
#STD# "|"		: {RETURN 31;}
#STD# "}"		: {RETURN 32;}
#STD# "~"		: {RETURN 33;}

#STD# AND		: {RETURN 34;}
#STD# ARRAY		: {RETURN 35;}
#STD# BEGIN		: {RETURN 36;}
#STD# BY		: {RETURN 37;}
#STD# CASE		: {RETURN 38;}
#STD# CONST		: {RETURN 39;}
#STD# DEFINITION	: {RETURN 40;}
#STD# DIV		: {RETURN 41;}
#STD# DO		: {RETURN 42;}
#STD# ELSE		: {RETURN 43;}
#STD# ELSIF		: {RETURN 44;}
#STD# END		: {RETURN 45;}
#STD# EXIT		: {RETURN 46;}
#STD# EXPORT		: {RETURN 47;}
#STD# FOR		: {RETURN 48;}
#STD# FROM		: {RETURN 49;}
#STD# IF		: {RETURN 50;}
#STD# IMPLEMENTATION	: {RETURN 51;}
#STD# IMPORT		: {RETURN 52;}
#STD# IN		: {RETURN 53;}
#STD# LOOP		: {RETURN 54;}
#STD# MOD		: {RETURN 55;}
#STD# MODULE		: {RETURN 56;}
#STD# \NOT              : {RETURN 57;}
#STD# OF		: {RETURN 58;}
#STD# OR		: {RETURN 59;}
#STD# POINTER		: {RETURN 60;}
#STD# PROCEDURE		: {RETURN 61;}
#STD# QUALIFIED		: {RETURN 62;}
#STD# RECORD		: {RETURN 63;}
#STD# REPEAT		: {RETURN 64;}
#STD# RETURN		: {RETURN 65;}
#STD# SET		: {RETURN 66;}
#STD# THEN		: {RETURN 67;}
#STD# TO		: {RETURN 68;}
#STD# TYPE		: {RETURN 69;}
#STD# UNTIL		: {RETURN 70;}
#STD# VAR		: {RETURN 71;}
#STD# WHILE		: {RETURN 72;}
#STD# WITH		: {RETURN 73;}

#STD# letter (letter | digit) *	: {RETURN 74;}

File: rex/test/modula/modula.cxx
GLOBAL {
# define yyInitFileStackSize 0

void Scanner::ErrorAttribute (int Token, tScanAttribute * Attribute)
{
   output ('c');
   unput ('a');
   yyEcho;
   yyEol (0);
   yyLess (1);
   yyPrevious;
   yyStart (STD);
   yyTab1 (0);
   yyTab2 (1, 2);
   yyTab;
}

/*
# define yySetPosition Attribute.Position.Line   = yyLineCount; \
  Attribute.Position.Column = (int) ((unsigned char *) TokenPtr - yyLineStart);\
  printf ("%ld\n", yyOffset (TokenPtr));
*/
}

DEFAULT {
   (void) printf ("%ld, %d: illegal character: ",
      Attribute.Position.Line, Attribute.Position.Column);
   yyEcho;
   (void) printf ("\n");
}

EOF {
if (yyStartState == Comment) (void) printf ("unclosed comment\n");
}

DEFINE
   digit	= {0-9}		.
   letter	= {a-z A-Z}	.
   CmtCh	= - {*(\t\n}	.

START Comment

RULE

#STD, Comment# "(*"	:- {yyPush (Comment);}
#Comment#  "*)"		:- {yyPop ();}
#Comment#  "(" | "*" | CmtCh + :- {}

#STD# digit +		,
#STD# digit + / ".."	: {return 1;}
#STD# {0-7} + B		: {return 2;}
#STD# {0-7} + C		: {return 3;}
#STD# digit {0-9 A-F} * H : {return 4;}
#STD# digit + "." digit * (E {+\-} ? digit +) ?	: {return 5;}

#STD# ' - {\n'} * '	|
      \" - {\n"} * \"	: {return 6;}

#STD# "#"		: {return 7;}
#STD# "&"		: {return 8;}
#STD# "("		: {return 9;}
#STD# ")"		: {return 10;}
#STD# "*"		: {return 11;}
#STD# "+"		: {return 12;}
#STD# ","		: {return 13;}
#STD# "-"		: {return 14;}
#STD# "."		: {return 15;}
#STD# ".."		: {return 16;}
#STD# "/"		: {return 17;}
#STD# ":"		: {return 18;}
#STD# ":="		: {return 19;}
#STD# ";"		: {return 20;}
#STD# "<"		: {return 21;}
#STD# "<="		: {return 22;}
#STD# "<>"		: {return 23;}
#STD# "="		: {return 24;}
#STD# ">"		: {return 25;}
#STD# ">="		: {return 26;}
#STD# "["		: {return 27;}
#STD# "]"		: {return 28;}
#STD# "^"		: {return 29;}
#STD# "{"		: {return 30;}
#STD# "|"		: {return 31;}
#STD# "}"		: {return 32;}
#STD# "~"		: {return 33;}

#STD# AND		: {return 34;}
#STD# ARRAY		: {return 35;}
#STD# BEGIN		: {return 36;}
#STD# BY		: {return 37;}
#STD# CASE		: {return 38;}
#STD# CONST		: {return 39;}
#STD# DEFINITION	: {return 40;}
#STD# DIV		: {return 41;}
#STD# DO		: {return 42;}
#STD# ELSE		: {return 43;}
#STD# ELSIF		: {return 44;}
#STD# END		: {return 45;}
#STD# EXIT		: {return 46;}
#STD# EXPORT		: {return 47;}
#STD# FOR		: {return 48;}
#STD# FROM		: {return 49;}
#STD# IF		: {return 50;}
#STD# IMPLEMENTATION	: {return 51;}
#STD# IMPORT		: {return 52;}
#STD# IN		: {return 53;}
#STD# LOOP		: {return 54;}
#STD# MOD		: {return 55;}
#STD# MODULE		: {return 56;}
#STD# \NOT              : {return 57;}
#STD# OF		: {return 58;}
#STD# OR		: {return 59;}
#STD# POINTER		: {return 60;}
#STD# PROCEDURE		: {return 61;}
#STD# QUALIFIED		: {
   output ('c');
   unput ('a');
   yyEcho;
   yyEol (0);
   yyLess (1);
   yyPrevious;
   yyStart (STD);
   yyTab1 (0);
   yyTab2 (1, 2);
   yyTab;
   return 62;}
#STD# RECORD		: {return 63;}
#STD# REPEAT		: {return 64;}
#STD# RETURN		: {return 65;}
#STD# SET		: {return 66;}
#STD# THEN		: {return 67;}
#STD# TO		: {return 68;}
#STD# TYPE		: {return 69;}
#STD# UNTIL		: {return 70;}
#STD# VAR		: {return 71;}
#STD# WHILE		: {return 72;}
#STD# WITH		: {return 73;}

#STD# letter (letter | digit) *	: {return 74;}
